zilliqa/tests/it/consensus.rs

```diff
@@ -1,11 +1,138 @@
+                    .get_finalized_height()
+                    >= start_block
+            },
+            50,
+        )
+        .await
+        .unwrap();
+
+    // Add a new node, then make sure they eventually are the producer of
+    // a block
+    info!("Adding node to network");
+    let index = network.add_node();
+
+    network
+        .run_until(
+            |n| {
+                let block = n.node_at(index)
+                        .get_latest_block()
+                        .unwrap()
+                        .unwrap();
+
+                    block.committee.leader_index(block.number()) == index
+            },
+            50000,
+        )
+        .await
+        .expect("expected to find our joined node has become a block producer");
+}
+*/
+
+//fn network_can_die_restart() {}
+
+// Test that all nodes can die and the network can restart (even if they startup at different
+// times)
+#[zilliqa_macros::test]
+async fn network_can_die_restart(mut network: Network) {
+    let start_block = 5;
+    let finish_block = 10;
+
+    let samples: usize = std::env::var_os("ZQ_TEST_SAMPLES")
+        .map(|s| s.to_str().unwrap_or("1").parse().unwrap_or(1))
+        .unwrap_or(1);
+
+    println!("{}", samples);
+
+    // wait until at least 5 blocks have been produced
+    network
+        .run_until(
+            |n| {
+                let index = n.random_index();
+                n.get_node(index).get_finalized_height() >= start_block
+            },
+            50,
+        )
+        .await
+        .unwrap();
+
+    // Forcibly restart the network, with a random time delay between each node
+    network.restart();
+
+    // Panic if it can't progress to the target block
+    network
+        .run_until(
+            |n| {
+                let index = n.random_index();
+                n.get_node(index).get_finalized_height() >= finish_block
+            },
+            50000,
+        )
+        .await
+        .expect("Failed to progress to target block");
+}
+
+fn get_block_number(n: &mut Network) -> u64 {
+    let index = n.random_index();
+    n.get_node(index).get_finalized_height()
+}
+
+// test that even with some consensus messages being dropped, the network can still proceed
+// note: this drops all messages, not just consensus messages, but there should only be
+// consensus messages in the network anyway
+#[zilliqa_macros::test]
+async fn block_production_even_when_lossy_network(mut network: Network) {
+    let failure_rate = 0.1;
+    let start_block = 5;
+    let finish_block = 8;
+
+    // wait until at least 5 blocks have been produced
+    network
+        .run_until(
+            |n| {
+                let index = n.random_index();
+                n.get_node(index).get_finalized_height() >= start_block
+            },
+            50,
+        )
+        .await
+        .unwrap();
+
+    // now, wait until block 15 has been produced, but dropping 10% of the messages.
+    for _i in 0..1000000 {
```

Comment:
```suggestion
    for _ in 0..1000000 {
```

Also, would a more reasonable timeout make sense here? If I'm reading this correctly the condition is just that the network advances a single block, I think a million messages is quite unreasonable even under unreliable conditions. I'd expect it should probably work within 50-100 messages comfortably, given that it works at all in the first place.

Reply:
yes, you're right, I'll lower it